### PR TITLE
Update instrumentation with Caliper

### PR DESCRIPTION
**Pull request by pramodk**

### CODE REVIEW

File: coreneuron/nrniv/main1.cpp [hunk 1]
```diff
@@ -396,7 +399,10 @@ extern "C" int run_solve_core(int argc, char** argv) {
     }
 
     // initializationa and loading functions moved to separate
+    Instrumentor::phase_begin("load-model");
     nrn_init_and_load_data(argc, argv, configs.size() > 0);
```
Comment (tristan0x) [inline coreneuron/nrniv/main1.cpp]: s/configs.size() > 0/!config.empty()/

File: CMakeLists.txt [hunk 2]
```diff
@@ -183,6 +184,18 @@ if(ENABLE_REPORTINGLIB)
     endif(reportinglib_FOUND)
 endif(ENABLE_REPORTINGLIB)
 
+# Code instrumentation
+if(ENABLE_CALIPER)
+    find_package(caliper REQUIRED)
+    include_directories(${caliper_INCLUDE_DIR})
+    add_definitions("-DCALIPER")
```
Comment (tristan0x) [inline CMakeLists.txt]: I would use a `CORENEURON_CALIPER` macro instead of `CALIPER`

File: coreneuron/nrnoc/fadvance_core.cpp [hunk 3]
```diff
@@ -34,6 +34,9 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include "coreneuron/nrniv/nrn_acc_manager.h"
 #include "coreneuron/utils/reports/nrnreport.h"
 #include "coreneuron/utils/progressbar/progressbar.h"
+#include "coreneuron/nrniv/profiler_interface.h"
+
+#include <sstream>
```
Comment (tristan0x) [inline coreneuron/nrnoc/fadvance_core.cpp]: standard include should be before coreneuron ones. clang-format should do it automatically.

**Comment by tristan0x** [comment]
If I am not mistaken, most of the end statements are in the same scope than the begin. So what an API that would leverages scopes like `scoped_lock` for instance:

Instead of:
```cpp
if (nrn_have_gaps) {
    Instrumentor::phase_begin("gap-v-transfer");
    nrnthread_v_transfer(_nt);
    Instrumentor::phase_end("gap-v-transfer");
}
```

have something like:
```cpp
if (nrn_have_gaps) {
    Instrumentor::phase p("gap-v-transfer");
    nrnthread_v_transfer(_nt);
}
```

**Comment by pramodk** [comment]
> If I am not mistaken, most of the end statements are in the same scope than the corresponding begin. So what an API that would leverages scopes like `scoped_lock` for instance:
> 
> Instead of:
> 
> ```c++
> if (nrn_have_gaps) {
>     Instrumentor::phase_begin("gap-v-transfer");
>     nrnthread_v_transfer(_nt);
>     Instrumentor::phase_end("gap-v-transfer");
> }
> ```
> 
> have something like:
> 
> ```c++
> if (nrn_have_gaps) {
>     Instrumentor::phase p("gap-v-transfer");
>     nrnthread_v_transfer(_nt);
> }
> ```

yes, c++ example of caliper works like this : https://llnl.github.io/Caliper/AnnotationAPI.html#function-annotations

I think we need timer & profiler interface that we can use across different codes. That more generic implementation can be tackled in separate PR ?

I have addressed rest of the comments.

**Comment by tristan0x** [comment]
> I think we need timer & profiler interface that we can use across different codes. That more generic implementation can be tackled in separate PR ?

Yes sure. No problem. I will create an issue about that to discuss it.

**Pull request merged**